Blog - blog post for Arcus.Messaging v0.4-v0.5 (#95)
arcus-azure/arcus: community/blog/2020.10-Introducing-message-handling-with-Azure-Service-Bus-message-influence/README.md

> # Introducing message handling with Azure Service Bus message influence in v0.5

Reviewer: It might be me, but when I read the title, I have no idea what the paragraph/chapter will talk about ?  What is meant with 'message influence' ?

Author: Yes, you may be right. I mostly took that from the [original issue](https://github.com/arcus-azure/arcus.messaging/issues/118) that used these words.
It's about calling the specific Azure Service Bus message operations during the message handling. Could you give some alternative, then? "Control Azure Service Bus message during message handling?"

Author: @tomkerkhove , should I change the title here? Any suggestions?

Starting from v0.4 and updated in v0.5, the [Arcus.Messaging](https://github.com/arcus-azure/arcus.messaging) has got an upgrade in handling messages from Azure Service Bus.
Previous versions already had set up a message handling system where messages could be routed and processed according to contextual information, but the system wasn't yet mature enough that it provided a way to report back to Azure when the message processing was completed.

Until now.

> ## Azure Service Bus message operations

Reviewer: I think you're 'deep diving' in this paragraph already, but I -as the reader who has no previous knowledge on Arcus messaging' am already lost.  Might be helpful to have some short high level intro on what Arcus Messaging is or tries to achieve.  (Or, if it is not possible to do that in a few sentences, link to a previous blogpost where Arcus Messaging is introduced ? )

Author: Yes, thanks. I added a paragragh with a link to our announcement post.

Reviewer 2: That's a good start indeed @stijnmoreels but we'll need to improve our docs a bit more then I think.
```suggestion
## Providing capability to directly decide what to do with Azure Service Bus messages
```

{
    public async Task ProcessMessageAsync(Message orderMessage, AzureServiceBusMessageContext context, ...)
    {
        // Processing order.
    }
}
```

When you decide you want to dead letter the message, you'll have to implement  `AzureServiceBusFallbackMessageHandler` instead.

```csharp
public class OrderFallbackMessageHandler : AzureServiceFallbackMessageHandler
{
    public override async Task ProcessMessageAsync(Message orderMessage, AzureServiceBusMessageContext context, ...)
    {
        // Processing order.
        await base.DeadLetterAsync(orderMessage);
    }
}
```

Note that because the fallback message handlers have more control over the original Azure Service Bus message, they will have to pass along the message to the base operations.

## Is that all?

No, of course not. The v0.4 and v0.5 version also has some minor updates regarding less verbose logging and some extra correlation extensions to pass along your own custom operation/transation ID property. All are helping in making the developer and consumer more in control of our messaging functionality.

## What's on the horizon?

> We'll see how we can reuse our message handling system in the context of Azure functions where the function binding will take the place of the message pump.

Reviewer: ```suggestion
We'll see how we can reuse our message handling system in the context of Azure functions where the function binding will replace the message pump.
```

We'll also look at how we can let the consumer be more in control of how the incoming Azure Service Bus messages are being deserialized before they're passed to the message handling.

> Have a question, remark, feedback or idea for a superdub addition?

Reviewer: ```suggestion
Have a question, remark, feedback or idea for a superduper addition?
```
?

> Feel free to take a look at our GitHub repository and create an issue.

Reviewer: Here, maybe link to the repo ?

Author: Yes, added that. Thanks. Also added the link to create an issue.

# Introducing message handling with Azure Service Bus message influence in v0.5

> Starting from v0.4 and updated in v0.5, the [Arcus.Messaging]() has got an upgrade in handling messages from Azure Service Bus.

Reviewer 2: ```suggestion
Starting from v0.4 and updated in v0.5, the [Arcus.Messaging](https://github.com/arcus-azure/arcus.messaging) has got an upgrade in handling messages from Azure Service Bus.
```

Author: Yeah, was planning to add it. 😄 . Thanks.

```csharp
public class OrderMessageHandler<Order> : AzureServiceBusMessageHandler<Order>
{
    public override async Task ProcessMessageAsync(Order order, AzureServiceBusMessageContext context, ...)
    {
        if (order.Customer is null)
        {
            // Calling base operation where the Azure Service Bus operations are located.
            await base.AbandonMessageAsync();
        }
        else
        {
            // Processing order.
        }
    }
}
```

Note that in the regular message handling you don't have to specify which message to abandon. This is all done behind the screens so the message handler only has to focus on the message processing and call the specific Azure Service Bus message operations when necessary.

### Fallback message handling

The fallback message handling is possible more correct when we're talking about dead lettering a message. Fallback message handlers will process the Azure Service Bus message when none other regular message handler was able to correctly process it.
At the end of the fallback message processing, the message could be dead lettered on Azure for example.

> Here's an example of how a normal `IServiceBusFallbackMessageHandler` is implemented:

Reviewer 2: Maybe show a registration example in this section how it has 2 message handlers and then this one as a fallback?

Author: Ah, yes, that would be good.

Author: Added some.

# Introducing message handling with Azure Service Bus message influence in v0.5

> [Arcus.Messaging]() has been arround quite a while now. It was created to remove the huge hurdle of constantly having to write the same boilerplate for processing message queues. 

Reviewer 2: ```suggestion
[Arcus Messaging](https://messaging.arcus-azure.net/) has been around for quite a while now. It was created to remove the huge hurdle of constantly having to write the same boilerplate for processing message queues. 
```

> See [our annoucing post](https://www.codit.eu/blog/announcing-arcus-messaging-background-jobs/) that introduced Arcus.Messaging for the first time for more information.

Reviewer 2: ```suggestion
On March 27th, 2020 we've [announced Arcus Messaging](https://www.codit.eu/blog/announcing-arcus-messaging-background-jobs/) for the first time which was just the start of a new journey to where we are today.
```

> Starting from v0.4 and updated in v0.5, the [Arcus.Messaging](https://github.com/arcus-azure/arcus.messaging) has got an upgrade in handling messages from Azure Service Bus.

Reviewer 2: ```suggestion
Starting from v0.4 and v0.5, the [Arcus Messaging](https://github.com/arcus-azure/arcus.messaging) has got an upgrade in how you can interact with messages coming from Azure Service Bus.
```

> Previous versions already had set up a message handling system where messages could be routed and processed according to contextual information, but the system wasn't yet mature enough that it provided a way to report back to Azure when the message processing was completed.

Reviewer 2: ```suggestion
Previous versions already had set up a message handling system where messages could be routed and processed according to contextual information, but it didn't cover all the scenarios that our customers have and we fell short, for example, it didn't provide a way to manually dead-letter a message.
```

Until now.

## Azure Service Bus message operations

> Currently, we provide three specific Azure Service Bus functions:

Reviewer 2: ```suggestion
We have introduced three new message operations for Azure Service Bus which you can now use in your message handlers:
```

- **Completing** a message which marks the Service Bus message 'complete' on Azure so no other queue can process it again (this is normally done automatically but can be turned off so you're in full control).
- **Dead lettering** a message which marks the Service Bus message 'dead lettered' on Azure so it's moved to the dead letter queue.
- **Abandoning** a message which marks the Service Bus message 'abandonned' on Azure.

## Specific Azure Service Bus message handlers

Because of the way the message handlers are registered and used inside the application, there's no 'callback' function to access the Azure Service Bus.

> That's why we've created two 'message handler templates' to help with this. These templates implements the message handler signature but also provides you with a whole range of specific Azure Service Bus functions that are related to completing the message processing.

Reviewer 2: What about this?
```suggestion

That's why we've created two dedicated message handlers for Azure Service Bus which allow you to interact closely with the message itself so that you can influence the message processing.
```

    {
        // Processing order.
    }
}
```

When you decide you want to abandon the message when the order is not recognized or the message context is invalid, you'll have to implement `AzureServiceBusMessageHandler<>` instead.

```csharp
public class OrderMessageHandler<Order> : AzureServiceBusMessageHandler<Order>
{
    public override async Task ProcessMessageAsync(Order order, AzureServiceBusMessageContext context, ...)
    {
        if (order.Customer is null)
        {
            // Calling base operation where the Azure Service Bus operations are located.
            await base.AbandonMessageAsync();
        }
        else
        {
            // Processing order.
        }
    }
}
```

Note that in the regular message handling you don't have to specify which message to abandon. This is all done behind the screens so the message handler only has to focus on the message processing and call the specific Azure Service Bus message operations when necessary.
The registration of this kind of message handler is just the same as any other regular message handler can can be added with the `.AddServiceBusMessageHandler` or `.AddMessageHandler` extensions.

> ### Fallback message handling

Reviewer 2: This is a new feature that we haven't discussed yet, maybe we should move it before this section and give it some more attention on the why and what it enables?

Author: Ah, you mean that we should explain more here what the fallback message handler does?

Reviewer 2: Jup

Author: OK, I added a small section above and referenced our docs where it's explained more deeply.

Reviewer 2: You mean you've added it here under `### Fallback message handling`, right?

I would move the introduction of the feature to an `##` as it's now buried under `## Specific Azure Service Bus message handlers`

Author: Aha, ok.

{
    public async Task ProcessMessageAsync(Message orderMessage, AzureServiceBusMessageContext context, ...)
    {
        // Processing order.
    }
}
```

When you decide you want to dead letter the message, you'll have to implement  `AzureServiceBusFallbackMessageHandler` instead.

```csharp
public class OrderFallbackMessageHandler : AzureServiceFallbackMessageHandler
{
    public override async Task ProcessMessageAsync(Message orderMessage, AzureServiceBusMessageContext context, ...)
    {
        // Processing order.
        await base.DeadLetterAsync(orderMessage);
    }
}
```

Note that because the fallback message handlers have more control over the original Azure Service Bus message, they will have to pass along the message to the base operations.

Registering this fallback message handler is just the same as when implementing the interface:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddServiceBusMessageHandler<OrderCompletedMessageHandler>()
            .AddServiceBusMessageHandler<OrderCancelledMessageHandler>()
            .AddServiceBusFallbackMessageHandler<OrderFallbackMessageHandler>();
}
```

## Is that all?

> No, of course not. The v0.4 and v0.5 version also has some minor updates regarding less verbose logging and some extra correlation extensions to pass along your own custom operation/transation ID property. All are helping in making the developer and consumer more in control of our messaging functionality.

Reviewer 2: ```suggestion
No, of course not!

We've released more improvements such as:
- Less verbose logging for our message pump and TCP health probes. All telemetry is still there, but moved to `Trace` verbosity
- Additional correlation extensions to pass along your own custom operation/transation ID property name

All are helping in making the developer and consumer more in control of our messaging functionality.
```

## What's on the horizon?

> We'll see how we can reuse our message handling system in the context of Azure functions where the function binding will replace the message pump.

Reviewer 2: ```suggestion
We are exploring how we can make our message handler approach available to customers using Azure Functions to provide a consistent experience. This would mean that Azure Functions would act as the message pump, but you can re-use your existing message handlers.
```

> We'll also look at how we can let the consumer be more in control of how the incoming Azure Service Bus messages are being deserialized before they're passed to the message handling.

Reviewer 2: ```suggestion

Next to that, we are working hard to give customers more control in how incoming messages are being deserialized before they're passed to the message handling.
```

> Have a question, remark, feedback or idea for a superduper addition?

Reviewer 2: ```suggestion
Have a question, remark, feedback or idea for a superduper addition? Feel free to take a look at our [GitHub repository](https://github.com/arcus-azure/arcus.messaging) and [create an issue](https://github.com/arcus-azure/arcus/issues/new/choose).
```

> Feel free to take a look at our [GitHub repository](https://github.com/arcus-azure/arcus.messaging) and [create an issue](https://github.com/arcus-azure/arcus/issues/new/choose).

Reviewer 2: Added to previous line
```suggestion
```

# Introducing message handling with Azure Service Bus message influence in v0.5

[Arcus Messaging](https://messaging.arcus-azure.net/) has been around for quite a while now. It was created to remove the huge hurdle of constantly having to write the same boilerplate for processing message queues. 
On March 27th, 2020 we've [announced Arcus Messaging](https://www.codit.eu/blog/announcing-arcus-messaging-background-jobs/) for the first time which was just the start of a new journey to where we are today.

Starting from v0.4 and v0.5, the [Arcus Messaging](https://github.com/arcus-azure/arcus.messaging) has got an upgrade in how you can interact with messages coming from Azure Service Bus.
Previous versions already had set up a message handling system where messages could be routed and processed according to contextual information, but it didn't cover all the scenarios that our customers have and we fell short, for example, it didn't provide a way to manually dead-letter a message.

Until now.

## Fallback message handling

> The capability to 'fallback' during message handling is also rather new. When a message is received on the message pump but there's not a single registered message handler that can handle the message correctly, the library provides a way to route your message to a fallback message handler.

Reviewer 2: ```suggestion
We've introduced the capability to fallback to a dedicated mesage handler when no other handler criteria is met.

When a message is received on the message pump and there is not a single registered message handler that can handle the message correctly based on the message routing, the library provides a way to route your message to a fallback message handler.
```

> This can of course be used when you expect that the received message will not always be handled. For more information on this topic, see our [dedicated docs page](https://messaging.arcus-azure.net/features/message-pumps/customization#fallback-message-handling) that explains this concept more thoroughly.

Reviewer 2: ```suggestion
This can be used as a fail-safe to create awareness of these messages or when you expect that the received message will not always be handled.

For more information on this topic, see our [dedicated docs page](https://messaging.arcus-azure.net/features/message-pumps/customization#fallback-message-handling) that explains this concept more thoroughly.
```

> The fallback message handling is possible more correct when we're talking about dead lettering a message. Fallback message handlers will process the Azure Service Bus message when none other regular message handler was able to correctly process it.

Reviewer 2: ```suggestion
The fallback message handling can be used to dead-letter messages, for example, so that an operator can have a look at them and decide they were not correctly processed due to a bug or if it's the expected behavior.
```

> At the end of the fallback message processing, the message could be dead lettered on Azure for example.

Reviewer 2: ```suggestion
```

Here's an example of how a normal `IServiceBusFallbackMessageHandler` is implemented.
`AzureServiceBusFallbackMessageHandler` is an `IServiceBusFallbackMessageHandler` implementation which provides the specific Azure Service Bus message operations

```csharp
public class OrderFallbackMessageHandler : IAzureServiceBusFallbackMessageHandler
{
    public async Task ProcessMessageAsync(Message orderMessage, AzureServiceBusMessageContext context, ...)
    {
        // Processing order.
    }
}
```

> When you decide you want to dead letter the message, you'll have to implement  `AzureServiceBusFallbackMessageHandler` instead.

Reviewer 2: ```suggestion
When you decide you want to dead-letter the message, you'll have to implement  `AzureServiceBusFallbackMessageHandler` instead.
```

> # Introducing message handling with Azure Service Bus message influence in v0.5

Reviewer 2: What about the suggestion above or here @fgheysels ?
```suggestion
# Introducing more message flow control with Azure Service Bus in Arcus Messaging v0.5
```

Reviewer: What about _Introducing improved message flow control with Azure Service Bus in Arcus Messaging_  @tomkerkhove ?

Reviewer 2: Sounds good to me!
